// ng2-contextmenu/bundles/contextmenu.umd.js
!function(e){function r(e,r,t){e in l||(l[e]={name:e,declarative:!0,deps:r,declare:t,normalizedDeps:r})}function t(e){return p[e]||(p[e]={name:e,dependencies:[],exports:{},importers:[]})}function n(r){if(!r.module){var o=r.module=t(r.name),a=r.module.exports,u=r.declare.call(e,function(e,r){if(o.locked=!0,"object"==typeof e)for(var t in e)a[t]=e[t];else a[e]=r;for(var n=0,u=o.importers.length;u>n;n++){var i=o.importers[n];if(!i.locked)for(var l=0;l<i.dependencies.length;++l)i.dependencies[l]===o&&i.setters[l](a)}return o.locked=!1,r},r.name);o.setters=u.setters,o.execute=u.execute;for(var s=0,d=r.normalizedDeps.length;d>s;s++){var f,c=r.normalizedDeps[s],v=l[c],m=p[c];m?f=m.exports:v&&!v.declarative?f=v.esModule:v?(n(v),m=v.module,f=m.exports):f=i(c),m&&m.importers?(m.importers.push(o),o.dependencies.push(m)):o.dependencies.push(null),o.setters[s]&&o.setters[s](f)}}}function o(r){var t={};if(("object"==typeof r||"function"==typeof r)&&r!==e)if(d)for(var n in r)"default"!==n&&a(t,r,n);else{var o=r&&r.hasOwnProperty;for(var n in r)"default"===n||o&&!r.hasOwnProperty(n)||(t[n]=r[n])}return t["default"]=r,c(t,"__useDefault",{value:!0}),t}function a(e,r,t){try{var n;(n=Object.getOwnPropertyDescriptor(r,t))&&c(e,t,n)}catch(o){return e[t]=r[t],!1}}function u(r,t){var n=l[r];if(n&&!n.evaluated&&n.declarative){t.push(r);for(var o=0,a=n.normalizedDeps.length;a>o;o++){var d=n.normalizedDeps[o];-1==s.call(t,d)&&(l[d]?u(d,t):i(d))}n.evaluated||(n.evaluated=!0,n.module.execute.call(e))}}function i(e){if(m[e])return m[e];if("@node/"==e.substr(0,6))return m[e]=o(v(e.substr(6)));var r=l[e];if(!r)throw"Module "+e+" not present.";return n(l[e]),u(e,[]),l[e]=void 0,r.declarative&&c(r.module.exports,"__esModule",{value:!0}),m[e]=r.declarative?r.module.exports:r.esModule}var l={},s=Array.prototype.indexOf||function(e){for(var r=0,t=this.length;t>r;r++)if(this[r]===e)return r;return-1},d=!0;try{Object.getOwnPropertyDescriptor({a:0},"a")}catch(f){d=!1}var c;!function(){try{Object.defineProperty({},"a",{})&&(c=Object.defineProperty)}catch(e){c=function(e,r,t){try{e[r]=t.value||t.get.call(e)}catch(n){}}}}();var p={},v="undefined"!=typeof System&&System._nodeRequire||"undefined"!=typeof require&&require.resolve&&"undefined"!=typeof process&&require,m={"@empty":{}};return function(e,t,n,a){return function(u){u(function(u){for(var l=0;l<t.length;l++)(function(e,r){r&&r.__esModule?m[e]=r:m[e]=o(r)})(t[l],arguments[l]);a({register:r});var s=i(e[0]);if(e.length>1)for(var l=1;l<e.length;l++)i(e[l]);return n?s["default"]:s})}}}("undefined"!=typeof self?self:global)

(["1"], ["3","5"], false, function($__System) {
var require = this.require, exports = this.exports, module = this.module;
$__System.register("2", ["3"], function (exports_1, context_1) {
    "use strict";
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var __moduleName = context_1 && context_1.id;
    var core_1, ContextmenuComponent, _a, _b;
    return {
        setters: [
            function (core_1_1) {
                core_1 = core_1_1;
            }
        ],
        execute: function () {
            ContextmenuComponent = (function () {
                function ContextmenuComponent(element, renderer) {
                    this.element = element;
                    this.renderer = renderer;
                    this.context = {};
                    this.isVisible = false;
                }
                ContextmenuComponent.prototype.show = function (x, y) {
                    this.renderer.setElementStyle(this.element.nativeElement, 'left', x + "px");
                    this.renderer.setElementStyle(this.element.nativeElement, 'top', y + "px");
                    this.isVisible = true;
                };
                ContextmenuComponent.prototype.hide = function () {
                    this.isVisible = false;
                };
                ContextmenuComponent.prototype.setContext = function (context) {
                    this.context = context;
                };
                ContextmenuComponent.prototype.get = function (path) {
                    return this.baseGet(this.context, path) || '';
                };
                // This function is borrowed from lodash
                // https://github.com/lodash/lodash/blob/master/lodash.js
                // Copyright JS Foundation and other contributors <https://js.foundation/>
                ContextmenuComponent.prototype.baseGet = function (object, path) {
                    path = Array.isArray(path) ? path : path.split('.');
                    var index = 0;
                    var length = path.length;
                    while (object != null && index < length) {
                        object = object[(path[index++])];
                    }
                    return (index && index === length) ? object : undefined;
                };
                ContextmenuComponent.prototype.onClick = function () {
                    this.hide();
                };
                ContextmenuComponent.prototype.onScroll = function () {
                    this.hide();
                };
                ;
                return ContextmenuComponent;
            }());
            __decorate([
                core_1.Input('id'),
                __metadata("design:type", String)
            ], ContextmenuComponent.prototype, "id", void 0);
            __decorate([
                core_1.HostListener('document:click', []),
                __metadata("design:type", Function),
                __metadata("design:paramtypes", []),
                __metadata("design:returntype", void 0)
            ], ContextmenuComponent.prototype, "onClick", null);
            __decorate([
                core_1.HostListener('document:scroll', []),
                __metadata("design:type", Function),
                __metadata("design:paramtypes", []),
                __metadata("design:returntype", void 0)
            ], ContextmenuComponent.prototype, "onScroll", null);
            ContextmenuComponent = __decorate([
                core_1.Component({
                    selector: 'contextmenu',
                    template: "\n    <div class=\"contextmenu\">\n      <ng-content></ng-content>\n    </div>\n  ",
                    styles: [
                        ":host(.contextmenu-container) {\n        display: none;\n        position: fixed;\n        z-index: 1300;\n    }",
                        ":host(.contextmenu-container.show) {\n        display: block;\n      }\n    "
                    ],
                    host: {
                        '[class.contextmenu-container]': '1',
                        '[class.show]': 'isVisible'
                    }
                }),
                __metadata("design:paramtypes", [typeof (_a = typeof core_1.ElementRef !== "undefined" && core_1.ElementRef) === "function" && _a || Object, typeof (_b = typeof core_1.Renderer !== "undefined" && core_1.Renderer) === "function" && _b || Object])
            ], ContextmenuComponent);
            exports_1("ContextmenuComponent", ContextmenuComponent);
        }
    };
});

$__System.register("4", ["3"], function (exports_1, context_1) {
    "use strict";
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var __moduleName = context_1 && context_1.id;
    var core_1, ContextmenuDirective;
    return {
        setters: [
            function (core_1_1) {
                core_1 = core_1_1;
            }
        ],
        execute: function () {
            ContextmenuDirective = (function () {
                function ContextmenuDirective() {
                }
                ContextmenuDirective.prototype.onContextmenu = function (ev) {
                    var _a = this.in, menu = _a[0], context = _a[1];
                    menu.setContext(context);
                    menu.show(ev.clientX, ev.clientY);
                    ev.stopPropagation();
                    ev.preventDefault();
                };
                return ContextmenuDirective;
            }());
            __decorate([
                core_1.Input('contextmenu'),
                __metadata("design:type", Array)
            ], ContextmenuDirective.prototype, "in", void 0);
            __decorate([
                core_1.HostListener('contextmenu', ['$event']),
                __metadata("design:type", Function),
                __metadata("design:paramtypes", [Object]),
                __metadata("design:returntype", void 0)
            ], ContextmenuDirective.prototype, "onContextmenu", null);
            ContextmenuDirective = __decorate([
                core_1.Directive({
                    selector: '[contextmenu]',
                    providers: []
                }),
                __metadata("design:paramtypes", [])
            ], ContextmenuDirective);
            exports_1("ContextmenuDirective", ContextmenuDirective);
        }
    };
});

$__System.register("1", ["3", "5", "2", "4"], function (exports_1, context_1) {
    "use strict";
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var __moduleName = context_1 && context_1.id;
    var core_1, platform_browser_1, contextmenu_component_1, contextmenu_directive_1, ContextmenuModule;
    return {
        setters: [
            function (core_1_1) {
                core_1 = core_1_1;
            },
            function (platform_browser_1_1) {
                platform_browser_1 = platform_browser_1_1;
            },
            function (contextmenu_component_1_1) {
                contextmenu_component_1 = contextmenu_component_1_1;
            },
            function (contextmenu_directive_1_1) {
                contextmenu_directive_1 = contextmenu_directive_1_1;
            }
        ],
        execute: function () {
            ContextmenuModule = (function () {
                function ContextmenuModule() {
                }
                return ContextmenuModule;
            }());
            ContextmenuModule = __decorate([
                core_1.NgModule({
                    imports: [
                        platform_browser_1.BrowserModule
                    ],
                    declarations: [
                        contextmenu_directive_1.ContextmenuDirective,
                        contextmenu_component_1.ContextmenuComponent
                    ],
                    bootstrap: [],
                    providers: [],
                    exports: [
                        contextmenu_directive_1.ContextmenuDirective,
                        contextmenu_component_1.ContextmenuComponent
                    ]
                }),
                __metadata("design:paramtypes", [])
            ], ContextmenuModule);
            exports_1("ContextmenuModule", ContextmenuModule);
        }
    };
});

})
(function(factory) {
  if (typeof define == 'function' && define.amd)
    define(["@angular/core","@angular/platform-browser"], factory);
  else if (typeof module == 'object' && module.exports && typeof require == 'function')
    module.exports = factory(require("@angular/core"), require("@angular/platform-browser"));
  else
    throw new Error("Module must be loaded as AMD or CommonJS");
});
